torch/nn/parallel/distributed.py

```diff
@@ -94,8 +94,19 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
         self.output_device = output_device
 
         # Sync params and buffers
-        for p in self.module.state_dict().values():
-            dist.broadcast(p, 0)
+
+        # Experimental code for NCCL backend
+        if dist._backend == "nccl":
+            grp = dist.new_group()
+            for p in self.module.state_dict().values():
+                dist.broadcast(p, 0, grp)
+            dist.barrier(grp)
+            dist.destroy_group(grp)
+            self.bcast_grp = dist.new_group()
+            self.reduce_grp = set()
```

Comment:
What's going on in here? Why do you create a temporary group, destroy it, and then create a brand new one? Can't you just reuse the same one? Also, why is `dist.barrier` necessary?

Reply:
Ok, but why? @csarofeen @ngimel any ideas? keeping the group alive would seem correct to me

Reply:
@csarofeen @ngimel Some further context, if grp is not destroyed here, distributed.py will deadlock after the next broadcast in sync_params() function

Reply:
@csarofeen @ngimel Wondering if you guys have any ideas why NCCL is deadlocking in this case? Each group creates the NCCL communicator and CUDA events and they will be cached. Destroy these groups will destroy both of the communicator and events.

if we don't destroy them, then after the first broadcast in distributed.py: __init__, after the next broadcast in distributed: _sync_params, it will deadlock. 

I'm happy to provide further contexts on the workaround